Add InsecureCookie permission provider role class

alexklbuckley opened the pull request:
- This will check if the new vufind_eds_guest cookie (defined in CookieConsent) is present in a users browsers cookies. (If cookie = 'vufind_eds_guest' is defined in the permissions.ini config file).
- If the vufind_eds_guest cookie is in the users browsers then it will assign the user the 'guest' role.

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley committed:
Add cookie permission provider role class

- This will check if the new vufind_eds_guest cookie (defined in
  CookieConsent) is present in a users browsers cookies. (If cookie =
'vufind_eds_guest' is defined in the permissions.ini config file).
- If the vufind_eds_guest cookie is in the users browsers then it will
  assign the user the 'guest' role.

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley committed:
Fix failing tests for adding cookie permission provider role class

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
Hi @demiankatz , 

I pushing my initial work for the cookie permission provider that you discussed on the VuFind tech mailing list with Craig Murdoch on June 12th. The emails were titled '[VuFind-Tech] [EXTERNAL] Guest access separate to logged in access'.

I've pushed a new Cookie.php permission provider class that checks for the presence of a cookie defined in the permissions.ini config - if it is present then the user is assigned 'guest' access.

Could I please get your thoughts on this? Any help much appreciated!

Kind regards,
Alex

demiankatz commented on config/vufind/CookieConsent.yaml:
```diff
@@ -100,6 +100,10 @@ CookieConsent:
           Domain: '{{vufind_cookie_domain}}'
           Description: CookieConsent::cookie_description_search_open_html
           Expiration: session
+        - Name: vufind_eds_guest
```
Rather than including an example in CookieConsent.yaml, it's probably better to leave this out (since details will vary by institution), and instead add documentation to permissions.ini, including a note reminding people to update CookieConsent.yaml with any custom cookies they add if they use the cookie consent feature.

alexklbuckley replied:
thanks @demiankatz I've included this in my follow-up patch!

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/Cookie.php:
```diff
@@ -0,0 +1,79 @@
+     *
+     * @param CookieManager $cookieManager Cookie manager
+     */
+    public function __construct(
```
Since this class requires a dependency to be injected, it looks like you need to create a factory and register it in the plugin manager. Those changes will need to be added to this PR before we can merge it.

alexklbuckley replied:
thanks @demiankatz i have added the CookieFactory and registered it with the plugin manager in my latest follow-up!

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/Cookie.php:
```diff
@@ -0,0 +1,79 @@
+ *
+ * PHP version 8
+ *
+ * Copyright (C) Villanova University 2007.
```
You might want to change the copyright institution and/or date.

alexklbuckley replied:
thanks @demiankatz I've included this in my follow-up patch!

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/Cookie.php:
```diff
@@ -0,0 +1,79 @@
+            }
+            $this->debug('getPermissions: result = true');
+        }
+        return ['guest'];
```
This should `return ['guest', 'loggedin'];` to make it more flexible. That way, used by itself, all roles will gain the permission when the cookie is set, but this can be filtered down by combining logic with the `Role` provider.

alexklbuckley replied:
thanks @demiankatz I've included this in my follow-up patch!

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/Cookie.php:
```diff
@@ -0,0 +1,79 @@
+    {
+        foreach ((array)$options as $cookieName) {
+            $this->debug("getPermissions: option cookieName '{$cookieName}'");
+            $cookie = $this->cookieManager->get($cookieName);
```
I wonder if we need some kind of mechanism for accounting for cookie values. If the cookie is just used as a flag to grant public access, then there's not really a security concern, but if we wanted a greater level of security, it might be better if we had some way of generating a random token in the session and then validating the cookie against that value, so people couldn't just edit their cookies to break in. That might be more work/complexity than is needed for your use case, and I certainly don't insist on over-engineering this... but if we keep it simple, we should include appropriate warnings in the documentation about any limitations related to ease of spoofing, etc.

alexklbuckley replied:
thanks @demiankatz I'm going to add these two remaining changes next!

alexklbuckley replied:
thanks @demiankatz we are thinking about if the simple solution is best for our use case or not. We'll get back to you shortly.

demiankatz reviewed (changes requested):
Thanks, @alexklbuckley, this is a great start! See below for some observations and suggestions.

alexklbuckley committed:
Cookie permission provider - implementing review improvements

- Remove specific 'vufind_eds_guest' from CookieConsent.yaml
- Add documentation to permissions.ini
- Change return value to return ['guest', 'loggedin'];
- Fix copyright institution name and date

Sponsored-by: Auckland University of Technology, New Zealand

demiankatz commented on config/vufind/permissions.ini:
```diff
@@ -48,6 +48,8 @@
 ; user        - Grant the permissions to logged in users whose user attribute match
 ;               the given regular expression pattern. For valid pattern syntax see
 ;               http://php.net/manual/de/reference.pcre.pattern.syntax.php.
+; cookie      - Grant the permissions to logged in users whose cookie value match the
```
The rest of this list is alphabetical, so it may be best to move this new entry to the top.

demiankatz replied:
I'm also not sure if the description is entirely accurate. Would this be closer to the current reality of what is happening?

```
Grant the permissions to users who have the named cookie(s) set in their browser.
```

As noted elsewhere, depending on if/how we incorporate value checks into the process, we will want to expand this description to either include information about how to specify values or else warnings about the security limitations of an approach based purely on cookie name.

alexklbuckley replied:
thanks @demiankatz I have re-ordered and reworded the description in the permissions.ini file in my latest follow-up.

more follow-ups to come.

demiankatz reviewed (changes requested):
Thanks for the progress, @alexklbuckley - I just left a couple of comments related to the permissions.ini comments. It may make sense not to act on them until after you've completed the latest version of the implementation, but I wanted to capture my thoughts before I forgot them.

alexklbuckley committed:
Cookie permission provider - implementing review improvements

- Create the CookieFactory
- Register CookieFactory with plugin manager
- Reorder and reword cookie documentation in permissions.ini

Sponsored-by: Auckland University of Technology, New Zealand

demiankatz committed:
Reduce diffs.

demiankatz commented on config/vufind/permissions.ini:
```diff
@@ -59,6 +61,7 @@
 ; ipRegEx = "/1\.2\.3\.4|1\.2\.3\.5/"
 ; ipRange[] = "1.2.3.4"
 ; ipRange[] = "1.2.3.7-1.2.5.254"
+; cookie = "custom cookie value"
```
Maybe this example should be something like "VUFIND_CUSTOM_COOKIE_NAME" so it's more realistic.

alexklbuckley replied:
thanks @demiankatz . I've added that in my latest follow-up commit

demiankatz reviewed (changes requested):
Thanks, @alexklbuckley -- one more suggestion on the .ini file. Also note that I just made a couple of small edits to the file to revert some minor accidental changes, just in the interest of having fewer overall diffs in the PR.

demiankatz committed:
Add missing alias; adjust indentation.

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/PluginManager.php:
```diff
@@ -46,6 +46,7 @@ class PluginManager extends \VuFind\ServiceManager\AbstractPluginManager
      * @var array
      */
     protected $aliases = [
+        'cookie' => Cookie::class,
```
I also added this missing alias for you. :-)

alexklbuckley replied:
thanks @demiankatz

alexklbuckley committed:
Merge branch 'vufind-org:dev' into cookie_permission_provider

alexklbuckley committed:
Improve the example cookie name in permissions.ini file

Sponsored-by: Auckland University of Technology, New Zealand

cmurdoch commented:
Hello @alexklbuckley and @demiankatz, from our point of view as the sponsor of this we'd like the simplest solution possible. So check the user cookie, and assign 'guest' permissions. We're obviously happy for you to advise on and alleviate any related security issues to the extent that's required by the community.

demiankatz commented:
> Hello @alexklbuckley and @demiankatz, from our point of view as the sponsor of this we'd like the simplest solution possible. So check the user cookie, and assign 'guest' permissions. We're obviously happy for you to advise on and alleviate any related security issues to the extent that's required by the community.

Here's a suggestion: if you want to keep this as-is, I think the code is acceptable, but it should be renamed `InsecureCookie` and the description should be changed to something like "Grant the permissions to users who have the named cookie(s) set in their browser. This can be easily spoofed by an end user, so it should not be used to protect secure features; however, it may be useful for setting flags to control unprotected behavior."

This then leaves the door open to implementing a SecureCookie subclass that adds extra layers of protection if we need that option in the future.

The simplest SecureCookie implementation I can think of would be to add a dependency on a `Laminas\Session\Container` object. This container could be keyed by cookie name and contain arbitrary values. The SecureCookie permission provider would not only check that the cookie is set, but also that its value matches the value in the session container. The idea is that the code that sets the cookie can also generate a random value and store it in the session container. This is not spoofable, since it requires the client and server to match on a value that cannot be predicted. It just requires a small amount of extra logic in the cookie-setting code to make it all work.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into cookie_permission_provider

alexklbuckley committed:
Add InsecureCookie permission provider class: Implement review feedback

- Rename Cookie.pm to InsecureCookie.pm

- Rename cookie permission to insecureCookie and alter the description

- Retain the CookieFactory.pm class

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
Thanks for the reply @demiankatz , much appreciated!

I've pushed a follow-up which implements what I hope are all your suggested changes for the InsecureCookie. Please let me know if any additional changes would be required?

You will notice I have kept the CookieFactory.pm name unchanged.

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/PluginManager.php:
```diff
@@ -46,6 +46,7 @@ class PluginManager extends \VuFind\ServiceManager\AbstractPluginManager
      * @var array
      */
     protected $aliases = [
+        'insecurecookie' => InsecureCookie::class,
```
This should probably be `insecureCookie` rather than `insecurecookie` for consistency.

alexklbuckley replied:
thanks @demiankatz I've implemented that in my latest follow up!

demiankatz commented on module/VuFind/src/VuFind/Role/PermissionProvider/CookieFactory.php:
```diff
@@ -0,0 +1,76 @@
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class CookieFactory implements \Laminas\ServiceManager\Factory\FactoryInterface
```
I would rename this to InsecureCookieFactory, since I'm pretty sure that if we add SecureCookie, it will have at least one additional dependency that would require a different factory.

alexklbuckley replied:
thanks @demiankatz I've implemented that in my latest follow up!

demiankatz reviewed (changes requested):
Thanks, @alexklbuckley! Just a couple of finishing touches suggested below.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into cookie_permission_provider

alexklbuckley committed:
Add InsecureCookie permission provider: Implement review feedback

- Fixing capitilisation of insecurecookie
- Rename CookieFactory.php to InsecureCookieFactory.php

Sponsored-by: Auckland University of Technology, New Zealand

demiankatz committed:
Add note about cookie consent.

demiankatz committed:
Add unit test.

demiankatz reviewed (approved):
Thanks, @alexklbuckley, I think this is ready to merge. 

Please take a look at my recent changes -- I added a small note to the .ini documentation reminding people to account for cookie consent, and I also added a unit test.

I'm approving this now, but will wait to hear from you before pushing the merge button, in case you have any comments on my additions.

alexklbuckley commented:
Thanks @demiankatz , your changes make perfect sense, thank you!

I have no further comments or questions on them :)

alexklbuckley commented:
thanks @demiankatz !
